File: app/src/main/java/com/simon/harmonichackernews/utils/DialogWindowUtils.java
package com.simon.harmonichackernews.utils;

import android.app.Dialog;
import android.content.Context;
import android.view.Window;
import android.view.WindowManager;

public class DialogWindowUtils {

    public static final int DEFAULT_MAX_WIDTH_DP = 500;

    private DialogWindowUtils() {
    }

    public static void applyMaxWidth(Dialog dialog) {
        applyMaxWidth(dialog, DEFAULT_MAX_WIDTH_DP);
    }

    public static void applyMaxWidth(Dialog dialog, int maxWidthDp) {
        if (dialog == null) {
            return;
        }

        Window window = dialog.getWindow();
        if (window == null) {
            return;
        }

        Context context = dialog.getContext();
        int maxWidthPx = Math.round(maxWidthDp * context.getResources().getDisplayMetrics().density);
        int horizontalMarginPx = Math.round(48 * context.getResources().getDisplayMetrics().density);
        int availableWidth = context.getResources().getDisplayMetrics().widthPixels - horizontalMarginPx;
        int targetWidth = Math.min(maxWidthPx, Math.max(0, availableWidth));
        if (targetWidth <= 0) {
            return;
        }

        window.setLayout(targetWidth, WindowManager.LayoutParams.WRAP_CONTENT);
    }
}

File: app/src/main/java/com/simon/harmonichackernews/utils/DialogWindowUtils.kt
package com.simon.harmonichackernews.utils

import android.app.Dialog
import android.view.WindowManager
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

object DialogWindowUtils {
    const val DEFAULT_MAX_WIDTH_DP = 500

    @JvmOverloads
    fun applyMaxWidth(dialog: Dialog?, maxWidthDp: Int = DEFAULT_MAX_WIDTH_DP) {
        val window = dialog?.window ?: return
        val displayMetrics = dialog.context.resources.displayMetrics
        val maxWidthPx = (maxWidthDp * displayMetrics.density).roundToInt()
        val horizontalMarginPx = (48 * displayMetrics.density).roundToInt()
        val availableWidth = displayMetrics.widthPixels - horizontalMarginPx
        val targetWidth = min(maxWidthPx, max(0, availableWidth))
        if (targetWidth <= 0) {
            return
        }

        window.setLayout(targetWidth, WindowManager.LayoutParams.WRAP_CONTENT)
    }
}
